### DISCUSSION TITLE
Connecting 2 peers. An example?

### DISCUSSION THREAD

**Discussion by ndrean** [elixir-webrtc/ex_webrtc#110] [Q&A] [answered]
I (try to) connect 2 peers to stream their camera once in a given room,  one on Chrome, the other of Firefox. Not there yet. The road seems long so I share some problems encountered. I would be glad to present a working example with your library.

I basically followed Reco with a Lobby that supervises and starts a dedicated Room GenServer (per room_id), and a JS hook (once in the room) that starts a  dedicated Channel per room. Perhaps only my JS is refactored a bit differently.

I encountered various errors and my process always dies.

- firstly the error below seems to be solved with the exclusion of self issued. I add a "user_id" to un-set_remote_description if the forwarded answer from Room is issued by the same peer. 

```
DOMException: Failed to execute 'setRemoteDescription' on 'RTCPeerConnection': Failed to set remote answer sdp: Called in wrong state: stable
```
- then I had :
```
DOMException: No remoteDescription.
```
which seemed fixed by queueing the candidates and adding them once the remoteDescription is set.

- if I start with Chrome, I get a few of these errors:

```
[error] Received data on failed pair! Ignoring. Pair id: 61320825457545405093851301635
```

- if I start with Firefox, I get:

```

DOMException: Answer has extmap http://www.ietf.org/id/draft-holmer-rmcat-transport-wide-cc-extensions-01 at level 0 that was not present in offer.
```

- If i use 2 tabs in Chrome, I end up with

```
[debug] EOC timer fired. Setting EOC flag.
[debug] Finished all conn checks, there won't be any further local or remote candidates
and we don't have any valid or selected pair. Changing connection state to failed.
```

When I look at [HANDLING THE INVITATION](https://developer.mozilla.org/en-US/docs/Web/API/WebRTC_API/Signaling_and_video_calling), they seem to follow exactly what `ex_Webrtc` does (or the opposite 😁). However, I don't see how the server can create an "answer" in response without having the stuff that I imagine the browser collects (such as codecs for the hardware present etc). Does this mean all the traffic goes through the server and not peer-to-peer?

**Comment by mickel8** [comment]
Hi @ndrean! Could you share your code with us? I know it's similar to reco but I would like to run it locally :)

**Reply by ndrean** [reply]
Sure: <https://github.com/ndrean/RTC>

**Reply by LVala** [reply, marked as answer]
Hi, I'm not sure if I follow, but your goal is to forward media from peer X to peer Y and from peer Y to peer X (so they see and hear each other, basically)? If so, it seems that you use only a single PeerConnection in your `Rtc.Room` module, when you should create a PeerConnection per peer.

PeerConnection is (simplifying a bit) a unique pair of IP addresses, so in your case, the setup would look like this: PeerConnection1 (Peer X <-> your server), PeerConnection2 (your server <-> Peer Y). Then you have to pass the media (so the RTP packets received from message `{:ex_webrtc, peer_connection_pid, {:rtp, track_id, packet}}`) from peer X to peer Y and peer Y to peer X.

Take a look at the [Broadcaster](https://github.com/elixir-webrtc/apps/tree/master/broadcaster/lib/broadcaster), especially at the modules `Forwarder` and `PeerSupervisor`. There, we create a single input PeerConnection and forward its media to a bunch of other output PeerConnections.

In `reco` this works as we assume one person per room (which, in retrospect, is a bit misleading) who sends his camera feed to the server and that's all. What you're trying to achieve is a bit more difficult.

It's worth mentioning that the negotiation of one PeerConnection (offer/answer exchange, ice candidates) is completely independent from other PeerConnections.

**Reply by LVala** [reply]
I'm not sure if that's what causes your issues, but that's the first thing I noticed.

**Reply by ndrean** [reply]
ah, starts to make much more sense. Yes, correct, forward media from peer X to peer Y and from peer Y to peer X. Will work on Broadcaster. Many thanks.

**Reply by mickel8** [reply]
Let us know when you encounter further problems as well as when you finish your app!
